### PR TITLE
feat: add a helm chart with the releaser action

### PR DESCRIPTION
**Pull request by allanger**
Hi!

I thought of starting to use `sql_exporter` on k8s, but I found out that there was no helm chart available, so I've decided to come up with one.

This PR contains the helm chart and the new GitHub action that will lint, test and release that chart. The release action requires GitHub pages enabled, because it's using it as a helm repository, you can check it in my fork: https://github.com/allanger/sql_exporter/tree/gh-pages

In this PR, release is enabled only when the action is triggered by pushing to the master, but since the action is checking if there were changes to the chart before releasing, it will require one more commit with a version bump to release the first version. 

When the new chart version is released, it will be put to GitHub releases and GitHub pages branch will be updated. Having it in GitHub releases may not be a desired behavior, since it will conflict with actual release.  If it's the case, I guess it's possible to disable it and put releases directly to the pages branch. 

What about maintenance?

I can take care of it for a time being. But basically, if there are no breaking changed introduced to the tool, the whole maintenance will be just upgrading `helm/Chart.yaml:appVersion` to the new app version and `helm/Chart.yaml/version` to the following one.

### CODE REVIEW

File: helm/values.yaml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+serviceAccount:
+  # Specifies whether a service account should be created
+  create: true
+  # Annotations to add to the service account
+  annotations: {}
+  # The name of the service account to use.
+  # If not set and create is true, a name is generated using the fullname template
+  name: ""
+test:
+  enabled: true
```
Comment (burningalchemist) [inline helm/values.yaml:22]: Is this parameter is resolved somewhere in the template? I didn't find it, but there might be some related logic (due to `helm test` command) behind the flag, that's why I'm asking. 👍

**Comment by burningalchemist** [comment]
Hey @allanger, thanks for your contribution! No objections here. The reason there's no helm chart is mostly people use `sql_exporter` as a container and put into their integrations. However, if there's a case to deploy it as a separate helm chart, then let it be. 👍

File: helm/values.yaml (hunk 1 above)
Comment (allanger) [reply helm/values.yaml:22]: It's not, I've forgotten to remove, I'll clean up

Comment (allanger) [reply helm/values.yaml]: It should be gone now

File: helm/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          - name: sql-exporter
+            readOnly: true
+            mountPath: /etc/sql_exporter/
+          {{- if .Values.collectorFiles}}
+          - name: sql-collector
+            readOnly: true
+            mountPath: /etc/sql_exporter/collectors/
+          {{- end }}
+          {{- if .Values.livenessProbe }}
+          livenessProbe: {{- toYaml .Values.livenessProbe | nindent 12 }}
```
Comment (burningalchemist) [inline helm/templates/deployment.yaml:58]: WDYT if we use `/healthz` for livenessProbe/readinessProbe of the container? I don't think users would put complicated logic behind `readinessProbe` (e.g. querying the actual backend from kubelet, this might be not the best idea), so we could just return HTTP 200 for both for starters.

In addition (just for context), there's a new feature coming, which would return prometheus output with `scrape_errors_total` counter instead of HTTP 500 (if target is named). Users then can configure this metric and check for scrape failures.

Comment (allanger) [reply helm/templates/deployment.yaml:58]: > WDYT if we use /healthz for livenessProbe/readinessProbe 

To be honest, I've copies this part from a PR to the old SQL_exporter repo. So yeah, I'll update them

Comment (allanger) [reply helm/templates/deployment.yaml:58]: > In addition (just for context), there's a new feature coming, which would return prometheus output with scrape_errors counter instead of HTTP 500 (if target is named). Users then can configure this metric and check for scrape failures.

That sounds nice, I can take care of tests once it's released

**Comment by tonobo** [comment]
Howdy everyone, it's possibly a bit out of context, but i'm interested in having a kubernetes controller. At least for handling the queries / metrics. Quite similar how prometheus rules work in the end. Is this a bit over the top for the project itself or would you like to see a contribution in there.

**Comment by burningalchemist** [comment]
Hey @tonobo, would you ind creating a separate discussion in `Discussions`? I would be curious to learn more about the idea, etc. 👍 It's just out of topic for this particular PR, still quite interesting. 🙂

File: helm/templates/NOTES.txt
[the remark this answers is not in the document]
Comment (allanger) [reply helm/templates/NOTES.txt:1]: Sure, I'll update tomorrow

File: helm/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    {{- include "sql-exporter.labels" . | nindent 4 }}
+spec:
+  replicas: {{ .Values.replicaCount }}
+  selector:
+    matchLabels:
+      {{- include "sql-exporter.selectorLabels" . | nindent 6 }}
+  template:
+    metadata:
+      annotations:
+        rollme: {{ randAlphaNum 5 | quote }}
```
Comment (burningalchemist) [inline helm/templates/deployment.yaml:19]: @allanger I think, in solutions like `ArgoCD` or `FluxCD` this line might cause constant resource drifts. I understand the intention, but I'm wondering if we can make it work on the `checksums` rather than random sequences. Otherwise, to drop the line and let it be populated with user-provided values?

File: helm/templates/NOTES.txt
[the remark this answers is not in the document]
Comment (burningalchemist) [reply helm/templates/NOTES.txt:1]: Looks good, thanks! 👏

File: helm/templates/deployment.yaml (hunk 3 above)
Comment (allanger) [reply helm/templates/deployment.yaml:19]: Ah, again, wasn't added by, copy-paste, I've overlooked. Let me check the whole chart one more time, I'll pay more attention to all the details. 
PS: I'm sorry, should've had a closer look before suggesting the PR, to get rid of all the strange stuff

Comment (burningalchemist) [reply helm/templates/deployment.yaml:19]: No worries, I also review file by file in my free time in the evening, so it goes slowly. We're not in a rush, though. 👍

Comment (allanger) [reply helm/templates/deployment.yaml]: I checked the whole chart again and fixed some stuff. Also, I've found out that ServiceMonitor was broken, so I've fixed it and added a test. That's why Values.tests are back, this tests is supposed to be triggered only in CI, so I'm enabling it in Values

File: helm/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v2
+name: sql-exporter
+description: Database agnostic SQL exporter for Prometheus
+type: application
+version: 0.1.0
+appVersion: 0.12.3
+keywords:
```
Comment (burningalchemist) [inline helm/Chart.yaml:7]: @allanger Would you mind adding yourself to the maintainers list? I'm referring to the reference field below:

```
maintainers: # (optional)
  - name: The maintainers name (required for each maintainer)
    email: The maintainers email (optional for each maintainer)
    url: A URL for the maintainer (optional for each maintainer)
```

File: helm/values.yaml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  {}
+  # capabilities:
+  #   drop:
+  #   - ALL
+  # readOnlyRootFilesystem: true
+  # runAsNonRoot: true
+  # runAsUser: 1000
+  # ---------------------------------------------------------------------
+  # -- Prometheus operator valuesnikolai.rodionov@grandcentrix.net
+  # ---------------------------------------------------------------------
```
Comment (burningalchemist) [inline helm/values.yaml:34]: @allanger  This part is a bit confusing to me. But I suggested adding your name in the maintainers list above. 👍

File: helm/values.yaml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+#         query: |
+#           SELECT Market, max(UpdateTime) AS LastUpdateTime
+#           FROM MarketPrices
+#           GROUP BY Market
+# ---------------------------------------------------------------------
+# -- These values only used for tests, users shouldn't set those
+# ---------------------------------------------------------------------
+tests:
+  serviceMonitor:
+    enabled: false
```
Comment (burningalchemist) [inline helm/values.yaml:111]: @allanger I'd say, we can safely drop this whole block just to avoid confusion for users (even though the comment says it all). So the defaults aren't set, but it's ok since the template should include respective resources only if `tests.serviceMonitor.enabled` is explicitly set. What do you think?

File: helm/values.yaml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  #                    COUNT(state)::float AS count
+  #                  FROM pg_stat_activity
+  #                  GROUP BY datname, usename, state;
+  #  collector_files:
+  #    - "*.collector.yml"
+
+# ---------------------------------------------------------------------
+# -- Collector Files
+# ---------------------------------------------------------------------
+# -- Collector files are mounter to /etc/sql_exporter/collectors dir
```
Comment (burningalchemist) [inline helm/values.yaml:86]: `mounter` typo. Sorry for nitpicks. 😅

File: helm/values.yaml (hunk 5 above)
Comment (allanger) [reply helm/values.yaml:34]: Damn, it was an accidental central mouse button click, I guess :)

File: helm/values.yaml (hunk 6 above)
Comment (allanger) [reply helm/values.yaml:111]: If we remove it from values, I'll have to add more complicated logic to templates, cause without this object, tests will become nil, and chart will not be installable at all, because the test template will be throwing panics. I can get rid of it and update templates to avoid it though, and you'll be able to say what you like more then

File: helm/values.yaml (hunk 7 above)
Comment (allanger) [reply helm/values.yaml:86]: No worries, I'll fix :)

File: helm/values.yaml (hunk 6 above)
Comment (burningalchemist) [reply helm/values.yaml:111]: @allanger Ah indeed! I've just checked out the PR, and bumped into nil stuff. 👍 
I applied a small change to the conditional wrapper like below:
```
{{- if (((.Values.tests).serviceMonitor).enabled)}}
...
{{- end }}
```
, and it seems to work just fine. I've also added the same wrapper in the `test-connection.yaml` file, otherwise it always gets rendered.

Please, let me know what you think, but to me it should be enough without spending extra time. I'm ready to merge the PR. 😃

Comment (allanger) [reply helm/values.yaml:111]: This condition doesn't seem a good option to filter out the `test-connection` test. And also, I thought that it would make sense to have test-connection always rendered, cause it can be used by users to see if their installation is set up correctly.

File: helm/values.yaml (hunk 7 above)
Comment (allanger) [reply helm/values.yaml]: It should be gone

File: helm/values.yaml (hunk 6 above)
Comment (allanger) [reply helm/values.yaml]: I've updated the serviceMonitor test

File: helm/values.yaml (hunk 5 above)
Comment (allanger) [reply helm/values.yaml]: It's gone

File: helm/Chart.yaml (hunk 4 above)
Comment (allanger) [reply helm/Chart.yaml:7]: I've added myself there

File: helm/values.yaml (hunk 6 above)
Comment (burningalchemist) [reply helm/values.yaml]: Yeah, agreed. Semantically this condition doesn't make sense for `test-connection` pod.

My concern is how useful is it to make it always rendered. Here's my point:

We have several cases when this test isn't indicative:

- Multiple connections (jobs): since targets are always named;
- Single named target.

We receive HTTP 200 with scrape duration, scrape errors and target up metrics for named targets. These values are to be scraped with Prometheus/AlertManager to react accordingly.

Situations when `sql_exporter` returns HTTP 500 are mostly for a single anonymous target:

- no metric families (`No metrics gathered`) - we don't return any of the above standard metrics, so there's literally nothing to show;
- no metrics encoded (`No metrics encoded`) - internal error due to the incorrect processing;

At the same time, I expect lots of errors in the beginning. Typically (based on the issues data), users install the exporter first, then start iterating to configure it. Most of the time they call the metric endpoint directly to check the data results until they're satisfied.

Comment (allanger) [reply helm/values.yaml]: I guess you're right. But I would use another condition to exclude it, wdyt?

Comment (burningalchemist) [reply helm/values.yaml]: Yeah, let's give it a shot. I don't have a strong opinion here, since I still learn (and am amazed at times) how people integrate the app in their workflows. 👍

Comment (allanger) [reply helm/values.yaml]: I've updated the condition, but I'm not so sure about naming anymore. Maybe you can come up with something better, or if it's fine with you, I guess the PR is done. The problem is that currently tests are not running because they are detecting changes to find out what to test, so I've triggered them by upgrading the Chart version here: https://github.com/allanger/sql_exporter/actions/runs/6448821034/job/17506501142

File: helm/values.yaml [hunk 8]
```diff
@@ -0,0 +1,107 @@
+# ---------------------------------------------------------------------
+# -- Global values
+# ---------------------------------------------------------------------
+image:
+  repository: burningalchemist/sql_exporter
+  pullPolicy:
+    IfNotPresent
```
Comment (burningalchemist) [inline helm/values.yaml:7]: I think it's a typo here (new line after `pullPolicy`).

File: helm/values.yaml (hunk 6 above)
Comment (burningalchemist) [reply helm/values.yaml]: I think it's all good. 🎉 We can handle the tests problem separately if needed. 👍

**Comment by burningalchemist** [comment]
@allanger Thanks again for your contribution! I guess we're good to go. 👍🚀

**Review by burningalchemist — APPROVED** [review]

**Comment by allanger** [comment]
If this PR is scheduled for a release with the new version, once the new version is released and the chart is merged, we could update fields `version` and `appVersion` in the `chart.yaml` then the chart should be released as well and become available from `gh-pages` branch.

**Comment by burningalchemist** [comment]
I've just noticed that the workflow is set to run on every `push` event. We might need to update the condition there to `tags` and/or `workflow_dispatch`:

```
on:
  workflow_dispatch:
  push:
    tags:
    - '*.*.*'
```

Could you please make this change?

**Comment by allanger** [comment]
It's triggered on every push because of the design. Chart doesn't have the same workflow as the app. Tools that are used for linting, testing, and releasing are detecting changes comparing to the master branch. So if the change to he chart is required, you don't need to push tags, but make a commit with an updated chart version to the master

**Comment by allanger** [comment]
It's triggered on every push because of the design. Chart doesn't have the same workflow as the app. Tools that are used for linting, testing, and releasing are detecting changes comparing to the master branch. So if the change to he chart is required, you don't need to push tags, but make a commit with an updated chart version to the master

**Comment by burningalchemist** [comment]
Ok, sounds good to me. :+1:
